File: src/app/android/app/src/main/java/ch/yarx/jasscardeye/ScoreRow.kt
package ch.yarx.jasscardeye

import android.os.Build
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.BlurredEdgeTreatment
import androidx.compose.ui.draw.blur
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.semantics.clearAndSetSemantics
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.res.pluralStringResource
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp

// Port of src/app/ios/Sources/ScoreRow.swift.

/**
 * The three-column score line, shared by the home screen and a running session so both read the same way:
 * own points left, what is being counted in the middle, the opponents' right.
 *
 * Each column takes exactly one third of the width, so the middle stays on the centre line and each number
 * is anchored to its own side while the digits change with every card.
 */
@Composable
fun ScoreRow(
    points: Int,
    pointsCaption: String,
    opponentPoints: Int,
    cards: Int,
    modeName: String,
    modifier: Modifier = Modifier,
    modeSuit: JassSuit? = null,
    note: String? = null,
    valueStyle: TextStyle = JassType.largeTitle,
    /** The demo: both numbers are drawn and blurred. The caller keeps the points out of the captions. */
    locked: Boolean = false,
) {
    val digits = valueStyle.copy(fontWeight = FontWeight.Bold, fontFeatureSettings = "tnum")
    Row(modifier = modifier.fillMaxWidth(), horizontalArrangement = Arrangement.spacedBy(8.dp)) {
        Column(Modifier.weight(1f), horizontalAlignment = Alignment.Start, verticalArrangement = Arrangement.spacedBy(1.dp)) {
            Value(points, digits, JassColors.Green, locked)
            Caption(pointsCaption)
        }
        Column(Modifier.weight(1f), horizontalAlignment = Alignment.CenterHorizontally, verticalArrangement = Arrangement.spacedBy(1.dp)) {
            Text(pluralStringResource(R.plurals.score_cards, cards, cards), style = JassType.callout.copy(fontFeatureSettings = "tnum"), color = JassColors.Secondary, maxLines = 1)
            Row(horizontalArrangement = Arrangement.spacedBy(4.dp), verticalAlignment = Alignment.CenterVertically) {
                if (modeSuit != null) SuitMark(modeSuit, 17.dp)
                Caption(modeName)
            }
            if (note != null) Caption(note)
        }
        Column(Modifier.weight(1f), horizontalAlignment = Alignment.End, verticalArrangement = Arrangement.spacedBy(1.dp)) {
            Value(opponentPoints, digits, JassColors.Orange, locked)
            Caption(stringResource(R.string.score_opponents))
        }
    }
}

/**
 * A score. Locked, it is drawn blurred - it still moves with every card, but cannot be read - and TalkBack hears that
 * it is locked instead of the number. Blur needs Android 12; below that the digits are drawn as dots of the same
 * count, which gives away no more.
 */
@Composable
private fun Value(number: Int, style: TextStyle, color: Color, locked: Boolean) {
    if (!locked) {
        Text("$number", style = style, color = color, maxLines = 1)
        return
    }
    val blurs = Build.VERSION.SDK_INT >= Build.VERSION_CODES.S
    val lockedLabel = stringResource(R.string.score_locked)
    Text(
        if (blurs) "$number" else "$number".replace(Regex("\\d"), "•"),
        style = style,
        color = color,
        maxLines = 1,
        modifier = Modifier
            .then(if (blurs) Modifier.blur(10.dp, BlurredEdgeTreatment.Unbounded) else Modifier)
            .clearAndSetSemantics { contentDescription = lockedLabel },
    )
}

/** Captions stay on one line rather than wrap - a column that grows a line would push the whole bar around. */
@Composable
private fun Caption(text: String) {
    Text(text, style = JassType.caption, color = JassColors.Secondary, maxLines = 1, overflow = TextOverflow.Ellipsis)
}

File: src/app/ios/Sources/ScoreRow.swift
import SwiftUI

/// The three-column score line, shared by the home screen and a running session so both read the
/// same way: own points left, what is being counted in the middle, the opponents' right.
///
/// Each column takes exactly one third of the width. That matters more than it looks: with plain
/// spacers the middle is only centred in whatever the two numbers leave over, so it drifts sideways
/// as the score grows from 0 to 1'764 - and during a count those numbers change with every card.
/// A third each keeps the middle on the centre line and anchors each number to its own side, so
/// nothing but the digits themselves moves.
struct ScoreRow: View {
    let points: Int
    /// Under the own score: "meine Punkte", or the breakdown "118 Karten +5 ×2".
    let pointsCaption: String
    let opponentPoints: Int
    let cards: Int
    let modeName: String
    /// The trump's mark, drawn next to its name. `nil` for a discipline without trump, which has
    /// no suit to show.
    var modeSuit: JassSuit? = nil
    /// Optional third middle line, e.g. "+5 letzter Stich".
    var note: String? = nil
    var valueFont: Font = .largeTitle
    /// The demo: both numbers are drawn and blurred, so they still move with every card
    /// but cannot be read. The caller keeps the points out of the captions.
    var locked = false

    var body: some View {
        HStack(alignment: .firstTextBaseline, spacing: 8) {
            column(alignment: .leading) {
                value(points, color: .green)
                caption(pointsCaption)
            }

            column(alignment: .center) {
                Text(String(localized: "score.cards", defaultValue: "\(cards) Karten"))
                    .font(.callout.monospacedDigit())
                    .foregroundStyle(.secondary)
                HStack(spacing: 4) {
                    if let modeSuit {
                        SuitMark(suit: modeSuit, size: 17)
                    }
                    caption(modeName)
                }
                if let note {
                    caption(note)
                }
            }

            column(alignment: .trailing) {
                value(opponentPoints, color: .orange)
                caption(String(localized: "score.opponents", defaultValue: "Gegner"))
            }
        }
    }

    /// A score. Locked, VoiceOver is told that it is locked rather than read the number the eye is
    /// not meant to see.
    private func value(_ number: Int, color: Color) -> some View {
        Text("\(number)")
            .font(valueFont.weight(.bold).monospacedDigit())
            .foregroundStyle(color)
            .blur(radius: locked ? 9 : 0)
            .accessibilityLabel(locked ? String(localized: "score.locked", defaultValue: "Punkte, freischalten") : "\(number)")
    }

    private func column<Content: View>(alignment: HorizontalAlignment,
                                       @ViewBuilder content: () -> Content) -> some View {
        VStack(alignment: alignment, spacing: 1) {
            content()
        }
        // One third each, so the middle sits on the centre line whatever the numbers do.
        .frame(maxWidth: .infinity, alignment: Alignment(horizontal: alignment, vertical: .center))
    }

    /// Captions stay on one line and shrink a little rather than wrap - a column that grows a line
    /// would push the whole bar around while counting.
    private func caption(_ text: String) -> some View {
        Text(text)
            .font(.caption)
            .foregroundStyle(.secondary)
            .lineLimit(1)
            .minimumScaleFactor(0.75)
    }
}
